executor.go

```diff
@@ -48,9 +52,190 @@ type ValidationResult struct {
+		Verifiers:      verifiers,
+		Store:          store,
+		PolicyEnforcer: policyEnforcer,
+	}, nil
+}
+
+// ValidateArtifact returns the result of verifying an artifact.
 func (e *Executor) ValidateArtifact(ctx context.Context, opts ValidateArtifactOptions) (*ValidationResult, error) {
-	return nil, errors.New("not implemented")
+	if err := validateExecutorSetup(e.Verifiers, e.Store); err != nil {
+		return nil, err
+	}
+
+	aggregatedVerifierReports, err := e.aggregateVerifierReports(ctx, opts)
+	if err != nil {
+		return nil, fmt.Errorf("failed to generate and aggregate verifier reports: %w", err)
+	}
+
+	if e.PolicyEnforcer == nil {
+		return &ValidationResult{
+			Succeeded:       false,
+			ArtifactReports: aggregatedVerifierReports,
+		}, nil
+	}
+
+	decision, err := e.PolicyEnforcer.Evaluate(ctx, aggregatedVerifierReports)
+	if err != nil {
+		return nil, fmt.Errorf("failed to evaluate verifier reports: %w", err)
+	}
+
+	return &ValidationResult{
+		Succeeded:       decision,
+		ArtifactReports: aggregatedVerifierReports,
+	}, nil
+}
+
+// aggregateVerifierReports generates and aggregates all verifier reports.
+func (e *Executor) aggregateVerifierReports(ctx context.Context, opts ValidateArtifactOptions) ([]*ValidationReport, error) {
+	var aggregatedVerifierReports []*ValidationReport
+
+	// Only resolve the root subject reference.
+	ref, err := e.resolveSubject(ctx, opts.Subject)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Implement a worker pool to validate artifacts concurrently.
+	// TODO: Enforce check on the stack size.
+	// Enqueue the subject artifact as the first task.
+	taskStack := stack.NewStack[*task]()
+	taskStack.Push(&task{
+		store:    e.Store,
+		artifact: ref.String(),
+		registry: ref.Registry,
+		repo:     ref.Repository,
+	})
+
+	for !taskStack.IsEmpty() {
+		// Ignore the bool as the stack is not empty.
+		task, _ := taskStack.Pop()
+
+		newTasks, artifactReports, err := e.verifySubjectAgainstReferrers(ctx, task, opts.ReferenceTypes)
+		if err != nil {
+			return nil, fmt.Errorf("failed to validate artifact %s: %w", task.artifact, err)
+		}
+
+		// Push the new tasks to the stack.
+		taskStack.Push(newTasks...)
+
+		// If the current task is the root task, add the artifactReports to the
+		// aggregatedVerifierReports. Otherwise, they are just artifactReports
+		// of the subject artifact.
+		if task.subjectReport != nil {
+			task.subjectReport.ArtifactReports = append(task.subjectReport.ArtifactReports, artifactReports...)
+		} else {
+			aggregatedVerifierReports = append(aggregatedVerifierReports, artifactReports...)
+		}
+	}
+
+	return aggregatedVerifierReports, nil
+}
+
+// verifySubjectAgainstReferrers verifies the subject artifact against all 
+// referrers in the store and produces new tasks for each referrer.
+func (e *Executor) verifySubjectAgainstReferrers(ctx context.Context, currentTask *task, referenceTypes []string) ([]*task, []*ValidationReport, error) {
+	var newTasks []*task
```

Comment:
nit: this line should go the around line 170. We only declare variables when we need it.